### PR TITLE
OEL-101: Europa Search Backend

### PR DESCRIPTION
**Pull request by saidatom**
### Description

Add a class EuropaSearchBackend extending Drupal\search_api\Backend\BackendPluginBase and implementing Drupal\Core\Plugin\PluginFormInterface

### Change log

- Added: EuropaSearchBackend with viewSettings, getSupportedFeatures, indexItems, deleteItems, deleteAllIndexItems, search, buildConfigurationForm
- Changed: N/A
- Deprecated: N/A
- Removed: N/A
- Fixed: N/A
- Security: N/A

### CODE REVIEW

File: src/Plugin/search_api/backend/EuropaSearchBackend.php [hunk 1]
```diff
@@ -0,0 +1,310 @@
...
+use Drupal\Core\StringTranslation\StringTranslationTrait;
+use Drupal\search_api\Backend\BackendPluginBase;
+use Drupal\search_api\IndexInterface;
+use Drupal\search_api\Query\QueryInterface;
+use OpenEuropa\EuropaSearchClient\Model\Document;
+use Solarium\QueryType\Select\Query\Query;
+use Symfony\Component\DependencyInjection\ContainerInterface;
+
+/**
+ * European Commission Europa Search backend for search_api.
```
Comment (drishu) [inline src/Plugin/search_api/backend/EuropaSearchBackend.php]:
```suggestion
 * Europa Search backend for search_api.
```

File: src/Plugin/search_api/backend/EuropaSearchBackend.php [hunk 2]
```diff
@@ -0,0 +1,310 @@
...
+
+  /**
+   * The config factory.
+   *
+   * @var \Drupal\Core\Config\ConfigFactoryInterface
+   */
+  protected $config;
+
+  /**
+   * The UserPlaces Helper custom module.
```
Comment (drishu) [inline src/Plugin/search_api/backend/EuropaSearchBackend.php]: Europa Search Service instance
or something of sorts

File: src/Plugin/search_api/backend/EuropaSearchBackend.php [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+    ];
+
+    $form['search_api_endpoint'] = [
+      '#type' => 'url',
+      '#title' => $this->t('Search API endpoint'),
+      '#description' => $this->t('The URL of the endpoint where the Search API is available.'),
+      '#required' => TRUE,
+      '#default_value' => $this->config->get('search_api_endpoint'),
+    ];
+
```
Comment (drishu) [inline src/Plugin/search_api/backend/EuropaSearchBackend.php]: lets add also 'token_api_endpoint' (available in develop branch)

File: src/Plugin/search_api/backend/EuropaSearchBackend.php [hunk 4]
```diff
@@ -0,0 +1,310 @@
...
+    return $form;
+  }
+
+  /**
+   * Form validation handler.
+   */
+  public function validateConfigurationForm(array &$form, FormStateInterface $form_state) {
+  }
```
Comment (drishu) [inline src/Plugin/search_api/backend/EuropaSearchBackend.php]: make sure the endpoints are actual urls

File: src/Plugin/search_api/backend/EuropaSearchBackend.php [hunk 5]
```diff
@@ -0,0 +1,310 @@
...
+
+  /**
+   * Form submission handler.
+   */
+  public function submitConfigurationForm(array &$form, FormStateInterface $form_state) {
+    $this->config
+      ->set('api_key', $form_state->getValue('api_key'))
+      ->set('database', $form_state->getValue('database'))
+      ->set('ingestion_api_endpoint', $form_state->getValue('ingestion_api_endpoint'))
+      ->set('search_api_endpoint', $form_state->getValue('search_api_endpoint'))
```
Comment (drishu) [inline src/Plugin/search_api/backend/EuropaSearchBackend.php]: also token_api_endpoint, same comment also for other references below

File: src/Plugin/search_api/backend/EuropaSearchBackend.php [hunk 6]
```diff
@@ -0,0 +1,310 @@
...
+    ];
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function indexItems(IndexInterface $index, array $items) {
+    $api = $this->client->ingestionApi();
+
+    // @todo Support multiple indexes by generating a reference id that takes
```
Comment (drishu) [inline src/Plugin/search_api/backend/EuropaSearchBackend.php]: If we dont do it now, let's create a follow up ticket and mark it here

File: src/Plugin/search_api/backend/EuropaSearchBackend.php [hunk 7]
```diff
@@ -0,0 +1,310 @@
...
+
+  /**
+   * {@inheritDoc}
+   */
+  public function search(QueryInterface $query) {
+    // @todo Make sure the search is run using the proper index.
+    $results = $this->client->searchApi();
+
+    $result_set = $query->getResults();
+    $result_set->setResultCount($results);
```
Comment (drishu) [inline src/Plugin/search_api/backend/EuropaSearchBackend.php]: the result count argument needs to be an int

File: src/Plugin/search_api/backend/EuropaSearchBackend.php [hunk 8]
```diff
@@ -0,0 +1,310 @@
...
+
+    foreach ($results as $document) {
+      $result_item = $this->fieldsHelper->createItem($query->getIndex(), $document->getReference());
+      $result_set->addResultItem($result_item);
```
Comment (drishu) [inline src/Plugin/search_api/backend/EuropaSearchBackend.php]: Fields need to be extracted

**Review by drishu — CHANGES_REQUESTED** [review]

File: src/Plugin/search_api/backend/EuropaSearchBackend.php (hunk 6 above)
Comment (saidatom) [reply src/Plugin/search_api/backend/EuropaSearchBackend.php]: follow up ticket OEL-149

File: src/Plugin/search_api/backend/EuropaSearchBackend.php (hunk 4 above)
Comment (saidatom) [reply src/Plugin/search_api/backend/EuropaSearchBackend.php]: remove this from new version

File: src/Plugin/search_api/backend/EuropaSearchBackend.php (hunk 7 above)
Comment (saidatom) [reply src/Plugin/search_api/backend/EuropaSearchBackend.php]: implemented $search->getTotalResults() in new file

**Pull request closed without merging**